lib/p2p/Pool.ts

```diff
@@ -185,14 +185,14 @@ class Pool extends EventEmitter {
       return;
     }
 
-    // ensure we stop listening for new peers before disconnecting from peers
+    // ensure we disconnect from existing peers before closing the server
```

Comment:
This change is required because otherwise the [server.close](https://nodejs.org/api/net.html#net_server_close_callback) triggered by `this.unlisten` never completed and resulted in an incomplete shutdown.

Reply:
Wouldn't the code above still result in a race condition between closing connections/peers and closing the server itself? I think it would still result in occasional incomplete shutdowns.

Reply:
I agree that it would be possible that we get a new connection after we've closed peers and pending connections, but before we stop listening. 

Ideally we'd have something like:
```
await server.pause()
await server.closeConnectionsAndPeers()
await server.close()
```

Reply:
@sangaman after reading the documentation of the [close event](https://nodejs.org/api/net.html#net_event_close) it looks like it'll always wait for the connections to be closed before emitting it. If I understand it correctly there's a guarantee that there won't be a race condition that I mentioned above.

That's a very elegant solution, thank you. :).

Reply:
Yeah so as I understand the callback for the `close()` method is simply listening for the `close` event, which is emitted when the server closes AND all connections are ended. The `close()` method stops accepting new connections, but doesn't close existing connections.

So to run through my thinking again we:

1) Call `close` up front, which stops accepting new connections ASAP.
2) Begin closing all existing sockets in parallel.
3) Wait for all sockets to close, and for the `close` event which should be emitted once both the server and all existing sockets are closed.